SECURITY.md

```diff
@@ -29,3 +35,89 @@ If you are editing this document please feel encouraged to change this format to
 - #993 Add restricted permissions to RocksDB - 1.4.5 - Anonymous security review
 - #740 Add hardening compile flags and `-fPIE` - 1.4.1 - Anonymous security review
 - #300 Add restricted permissions to osqueryd logging - 1.0.4
+
+## Security Requirements Met by Functionality
+
+### Confidentiality
+
+#### No Retrieval of Sensitive User-owned Data
+
+- Certain tables disabled by default, additional tables can be disabled
+
+#### Use of HTTPS (TLS)
+
+- HTTPS (specifically the TLS v1.2 protocol) is used to encrypt all communications
+
+#### No custom cryptography (only well-known libraries)
+
+#### No network connectivity to third parties
+
+#### Does not accept incoming network connections
+
+#### Does not 'shell out' to other binaries for data collection
+
+#### Access control functions
+
+### Integrity
+
+#### Limited attack surface
+
+#### Daemon runs as root
+
+- Its files are owned as root (config, RocksDB) to resist tampering
+- Extensions must also run as root, communicate over IPC restricted to root
+
+#### No self-update feature
+
+#### No methods to modify data on the system
+
+#### No dynamic linking
+
+#### Code-signed releases
+
+### Availability
```

Comment:
In security terms, the [Availability](https://en.wikipedia.org/wiki/Information_security#Key_concepts) pillar of the Confidentiality/Integrity/Availability triad. This header is the theme of the following subheaders.

Reply:
I think there's a place to talk about our security goals/principle.
* No network access (we're not perfect)
* read-only
* ???

I don't think I've heard about `availability` as one of our goals, though I can understand it as a pillar of system design. I think we have a big caveat about how ultimately, we're a root daemon. No kernel presence, and no real attempts at tamper proofing. But anyhow...

I sorta see this as shaping into 2 docs. One is our goals/principles. The other is more of a security questionnaire about how our principles map to some practical matter. I think there's some room to discuss, but my gut sense is these are dramatically different, and are at the least, different sections,

Reply:
Yea, I don't want this to have a checklist feel like we are covering topics only so that we've gone down some arbitrary list of things to say. So I agree, we should be intentional with what we put in here.

If I understand you, you're suggesting only including an enumeration of our claims, things we do protect, things we are doing. Whereas I was envisioning a candid document that anticipates and addresses the sort of criticisms a security auditor would have. "Oh boy you didn't think about THIS did you" or "what about the ..." and just being really candid in this document: this is in scope, this is fully mitigated, this is partially mitigated or is a best-effort thing, and we have decided this other thing is just out of scope.

Reply:
It's taken me a bit to get there, but I am suggesting we have 2 documents (or two sections). One is the goals/principles/tenets of osquery development, I imagine this is slightly philosophical. The other is a more pragmatic application of those, and is a bit closer to straight q&a style.

I think you're writing this closer to the security q&a style, but I'm getting tangled around what our principles are.

Reply:
Do the guidelines for contributions capture the tenents of development? For example https://github.com/osquery/osquery/blob/master/CONTRIBUTING.md#guidelines-for-contributing-features-to-osquery-core

Reply:
For the 'guiding principles' that are security relevant, I've elaborated on them here. I've also just added a cross-reference link to the CONTRIBUTING document.